cms/server/AdminWebServer.py

```diff
@@ -1675,81 +2061,93 @@ def post(self, contest_id):
     """Called to remove an announcement.
 
     """
-    def get(self, ann_id):
+    def get(self, contest_id, ann_id):
         ann = self.safe_get_item(Announcement, ann_id)
-        contest_id = ann.contest.id
+        self.contest = self.safe_get_item(Contest, contest_id)
+
+        # Additional check.
+        if self.contest is not ann.contest:
+            raise tornado.web.HTTPError(404)
 
         self.sql_session.delete(ann)
 
         try_commit(self.sql_session, self)
-        self.redirect("/announcements/%s" % contest_id)
+        self.redirect("/contest/%s/announcements" % contest_id)
 
 
-class UserViewHandler(BaseHandler):
-    """Shows the details of a single user (submissions, questions,
-    messages), and allows to send the latters.
+class ParticipationHandler(BaseHandler):
+    """Shows the details of a single user in a contest: submissions,
+    questions, messages (and allows to send the latters).
 
     """
-    def get(self, user_id):
-        user = self.safe_get_item(User, user_id)
-        self.contest = user.contest
+    def get(self, contest_id, user_id):
+        self.contest = self.safe_get_item(Contest, contest_id)
+        participation = self.sql_session.query(Participation)\
+                            .filter(Participation.contest_id == contest_id)\
+                            .filter(Participation.user_id == user_id)\
+                            .first()
+
+        # Check that the participation is valid.
+        if participation is None:
+            raise tornado.web.HTTPError(404)
 
         self.r_params = self.render_params()
-        self.r_params["selected_user"] = user
-        self.r_params["submissions"] = user.submissions
-        self.render("user.html", **self.r_params)
+        self.r_params["participation"] = participation
+        self.r_params["submissions"] = participation.submissions
+        self.r_params["selected_user"] = participation.user
+        self.render("participation.html", **self.r_params)
 
-    def post(self, user_id):
-        user = self.safe_get_item(User, user_id)
-        self.contest = user.contest
+    def post(self, contest_id, user_id):
+        fallback_page = "/contest/%s/user/%s" % (contest_id, user_id)
 
-        try:
-            attrs = user.get_attrs()
+        self.contest = self.safe_get_item(Contest, contest_id)
+        participation = self.sql_session.query(Participation)\
+                            .filter(Participation.contest_id == contest_id)\
+                            .filter(Participation.user_id == user_id)\
+                            .first()
 
-            self.get_string(attrs, "first_name")
-            self.get_string(attrs, "last_name")
-            self.get_string(attrs, "username", empty=None)
-            self.get_string(attrs, "password")
-            self.get_string(attrs, "email")
+        # Check that the participation is valid.
+        if participation is None:
+            raise tornado.web.HTTPError(404)
```

Comment:
Maybe you can save a few lines factoring out these in an _init method for this class.